# Transcriptome Analysis {#sec-transcript}

::: {.callout-important title="IMPORTANT" collapse=false}
**PLEASE NOTE** WE INCLUDED THE POSSIBILITY OF THE HUMAN TRANSCRIPTOME DATA ANALYSIS FROM THE SEQUENCE DATA. NONE OF THE OWNER OF THE CODE OR DATA CENTER IS RESPONSIBLE FOR HANDLING THE HUMAN DATA. IT IS UPTO THE USER IF THEY WANT TO USE THE ONLINE VERSION FOR HUMAN DATA ANALYSIS.
:::

Transcriptome analysis has divided in several parts for the data upload, analysis and visualizing the results.

## Data input and parameters settings
In this section, for better user accessibility, we divided this into 3 sub-parts -

![Fastq upload](images/transcriptome/transcript1.png){#fig-trans1}

### Fastq upload
1. If the user performed the ***Quality analysis*** (@sec-seqqual) already, they don't need to upload the data again, they can use ***'continue from Quality Analysis'***.
2. The user can also upload the data directly by selecting ***'upload data'*** option.

:::{.callout-note }
Please note it will take some time to upload the data and once the data uploaded, the list will be displayed in the **Data file viewer** (see @fig-qual2 @fig-trans1).
:::

### Sample info {#sec-sampleinfo}
In the ***Sample info*** section, users need to upload the 'sample sheet/meta data file' that includes atleast two columns ('sample name' and 'sample condition') (see @fig-trans2).
For Differential Gene Expression (DEG) analysis, user need to provide the comparison conditions as showed in @fig-trans2. 

![Sample information](images/transcriptome/transcriptr3.png){#fig-trans2}

:::{.callout-tip}
1. After uploading the files, all files will be appear in the browser window (see  @fig-trans2), please check and verify everything is properly uploaded.
2. User can upload any file format, such as text (.txt), CSV (.csv), Excel (.xls, .xlsx).
:::

### Parameters settings
Selecting the right parameters for the analysis is the key to the successful run of **TranscriptR**. The **parameters settings** (see @fig-trans3) are sub-divided for better understanding and view - 

![Parameters settings](images/transcriptome/transcriptr4.png){#fig-trans3}

#### Species and database
1. **Select Species**: The user need to select the *species* from the species list. Here we provided more than 350 species presented in the 'biomart' for the analysis. Note that all other species than 'Human', 'Mouse' and 'Rat' come with the **scientific name** as provided in ['Ensembl'](http://www.ensembl.org/info/about/species.html){target="_blank"}. The default is 'Human'.
2. **Select release version**: The default is *latest*. The user can choose other database version as provided by [Ensembl](https://ftp.ensembl.org/pub/){target="_blank"}. Please provide the number only in the 'box', example, if you want to use version '109', provide only '109' (see @fig-trans4).

![Parameters: Species and database](images/transcriptome/transcript5.png){#fig-trans4}

#### Alignment parameters
1. **Choose R1/R2 suffix**: For *paired-end* analysis, please provide the R1 and R2 suffixes.
2. **Overhang**: The default is 100. 
3. **Choose sequence type**: Choose whether the sequence is *stranded*, *unstranded* or *reverse-stranded*.
4. **Choose ID**: Select display ID for *gene id* or *transcriptome id*. Default is *gene ID*.

![Parameters: Alignment](images/transcriptome/transcript6.png){#fig-trans5}

#### DEG filtration
As **'TranscriptR'** provides the differential gene expression (DEG) analysis result, user need to provide the required filtration for the analysis. \
1. **Enter CPM value**: The default is 1. \
2. **Number of samples**: The default is 3. \
3. **Choose logFC value**: The default is 1.5. \
4. **Choose FDR value**: The default is 0.05.

![Parameters: DEF filtration](images/transcriptome/transcript7.png){#fig-trans6}

## Visualization: MultiQC
Please see @sec-seqqual for details.

## Visualization: Plots
Since **TranscriptR** provides differential gene expression (DEG) analysis by **three** different pipelines, namely *EdgeR*, *DESeq2* and *limma*, all plots are divided into these three results.

![Visualization: plots](images/transcriptome/transcript8.png){#fig-trans7}

:::{.callout-note}
1. The plots will be displayed as @fig-trans7 in a PDF format. All of these can be downloaded when the user download the results from the *'Download results'* tab.
2. Please note, if the plots are not loaded after the *job* completed and clicked on 'show BoxPlot/Show MDS Plot' button, refresh the browser and it will appear in the box. 
:::


## Visualization: DEG table
We provided differential gene expression (DEG) analysis result as a table format (see @fig-trans8). To display the table, user need to - \ 
1. choose the analysis type (as **TranscriptR** provides three different pipeline results), such as *EdgeR analysis*, *DESeq analysis* or *limma analysis* and \
2. select DEG table from the drop-down list. The list is generated from the information provided by the *comparison table*.

![Visualization: DEG table](images/transcriptome/transcript9.png){#fig-trans8}

In the DEG table, we added information for - \
1. **ID**: ENSEMBL *gene ID* or *transcript ID* \
2. **logFC**: log fold change value for each ID. Red color shows the higher expression and blue represents the low expression (same corresponds to *gene symbol*). \
3. **logCPM**: log CPM value provided \
4. **P_value**: *p*-value provided for each ID. \
5. **FDR**: FDR provided for each ID. \
6. **Gene Symbol**: official gene symbol added in the table. Coloured according to the *logFC* value. \
7. **Ensembl link**: we added external link for ensembl database to view further information on the ID. \
8. **GeneCard link**: we added genecard database link for more information on the ID. \
9. **Gene details**: Clicking on *'show details* button will result to display the boxplot of CPM data for each group (**see the note below**) and the gene information from **HGNC** (human only). \
10. **Chr**: Chromosomal location \
11. **Start**: Start position of the chromosome \
12. **End**: End position of the chromosome \
13. **Strand**: chromosomal strand \
14. **Types**: types of genes/transcript.

:::{.callout-important}
Please note, we already mentioned that the user may need to reload/refresh the browser after successful run to visualize the results. In this process, the uploaded data table (**mentioned in @sec-sampleinfo**) will be lost. To view the CPM boxplot, the user need the upload the sample information data tables again and click on *'Show details'* on the DEG table.
:::


## Visualization: InterSectDE
Further, we implemented intersection function for the above-mentioned three types of DEG analyses in the *'InterSectDEG* tab. \
1. To visualize the result, the user needs to select the *same result* table (eg. groupAvsgroupB) for all three analyses. **Please remember selecting different table may results incorrect table and figure. \
2. After selecting same result table files from all three analyses, click to *'show intersect result'*. Both intersect table and Venn diagram will appear in respective boxes. \
3. To download the full table, check to show **all** entries (see @fig-trans9). \
4. The Venn diagram can be downloaded in different file format with respective height, width or resolution (see @fig-trans9).

![InterSect result](images/transcriptome/transcript10.png){#fig-trans9}

Below is the details of *'InterSect DEG'* table result (see @fig-trans9) - \
1. **id** \
2. **name** \
3. **chr** \
4. **start** \
5. **end** \
6. **strand** \
7. **biotype** \
8. **edgeR_logFC** \
9. **edgeR_logCPM** \
10. **edgeR_PValue** \
11. **edgeR_FDR** \
12. **edgeR_baseMean** \
13. **DESeq2_log2FoldChange** \
14. **DESeq2_lfcSE** \
15. **DESeq2_stat** \
16. **DESeq2_pvalue** \
17. **DESeq2_padj** \
18. **limma_logFC** \
19. **limma_AveExpr** \
20. **limma_t** \
21. **limma_P.Value** \
22. **limma_adj.P.Val** \
23. **limma_B**

## Session & logs
We included the next tab to show *'GTF information'*, *'Reference information'*, *'Shiny logs'* and *'Session information'* for the pipeline run.


## Download Results
In the *'Download results'* tab, user can download all results performed and saved during the **Transcriptome Analysis** run in a zip folder. This also contains the snakemake and session logs.