<!DOCTYPE html>
<html xmlns="http://www.w3.org/1999/xhtml" lang="en" xml:lang="en"><head>

<meta charset="utf-8">
<meta name="generator" content="quarto-1.2.335">

<meta name="viewport" content="width=device-width, initial-scale=1.0, user-scalable=yes">


<title>TranscriptR manual - 2&nbsp; Transcriptome Analysis</title>
<style>
code{white-space: pre-wrap;}
span.smallcaps{font-variant: small-caps;}
div.columns{display: flex; gap: min(4vw, 1.5em);}
div.column{flex: auto; overflow-x: auto;}
div.hanging-indent{margin-left: 1.5em; text-indent: -1.5em;}
ul.task-list{list-style: none;}
ul.task-list li input[type="checkbox"] {
  width: 0.8em;
  margin: 0 0.8em 0.2em -1.6em;
  vertical-align: middle;
}
</style>


<script src="site_libs/quarto-nav/quarto-nav.js"></script>
<script src="site_libs/quarto-nav/headroom.min.js"></script>
<script src="site_libs/clipboard/clipboard.min.js"></script>
<script src="site_libs/quarto-search/autocomplete.umd.js"></script>
<script src="site_libs/quarto-search/fuse.min.js"></script>
<script src="site_libs/quarto-search/quarto-search.js"></script>
<meta name="quarto:offset" content="./">
<link href="./igv.html" rel="next">
<link href="./quality.html" rel="prev">
<script src="site_libs/quarto-html/quarto.js"></script>
<script src="site_libs/quarto-html/popper.min.js"></script>
<script src="site_libs/quarto-html/tippy.umd.min.js"></script>
<script src="site_libs/quarto-html/anchor.min.js"></script>
<link href="site_libs/quarto-html/tippy.css" rel="stylesheet">
<link href="site_libs/quarto-html/quarto-syntax-highlighting.css" rel="stylesheet" id="quarto-text-highlighting-styles">
<script src="site_libs/bootstrap/bootstrap.min.js"></script>
<link href="site_libs/bootstrap/bootstrap-icons.css" rel="stylesheet">
<link href="site_libs/bootstrap/bootstrap.min.css" rel="stylesheet" id="quarto-bootstrap" data-mode="light">
<script id="quarto-search-options" type="application/json">{
  "location": "sidebar",
  "copy-button": false,
  "collapse-after": 3,
  "panel-placement": "start",
  "type": "textbox",
  "limit": 20,
  "language": {
    "search-no-results-text": "No results",
    "search-matching-documents-text": "matching documents",
    "search-copy-link-title": "Copy link to search",
    "search-hide-matches-text": "Hide additional matches",
    "search-more-match-text": "more match in this document",
    "search-more-matches-text": "more matches in this document",
    "search-clear-button-title": "Clear",
    "search-detached-cancel-button-title": "Cancel",
    "search-submit-button-title": "Submit"
  }
}</script>


</head>

<body class="nav-sidebar floating">


<div id="quarto-search-results"></div>
  <header id="quarto-header" class="headroom fixed-top">
  <nav class="quarto-secondary-nav" data-bs-toggle="collapse" data-bs-target="#quarto-sidebar" aria-controls="quarto-sidebar" aria-expanded="false" aria-label="Toggle sidebar navigation" onclick="if (window.quartoToggleHeadroom) { window.quartoToggleHeadroom(); }">
    <div class="container-fluid d-flex justify-content-between">
      <h1 class="quarto-secondary-nav-title"><span class="chapter-number">2</span>&nbsp; <span class="chapter-title">Transcriptome Analysis</span></h1>
      <button type="button" class="quarto-btn-toggle btn" aria-label="Show secondary navigation">
        <i class="bi bi-chevron-right"></i>
      </button>
    </div>
  </nav>
</header>
<!-- content -->
<div id="quarto-content" class="quarto-container page-columns page-rows-contents page-layout-article">
<!-- sidebar -->
  <nav id="quarto-sidebar" class="sidebar collapse sidebar-navigation floating overflow-auto">
    <div class="pt-lg-2 mt-2 text-left sidebar-header">
    <div class="sidebar-title mb-0 py-0">
      <a href="./">TranscriptR manual</a> 
        <div class="sidebar-tools-main tools-wide">
    <a href="https://github.com/JD2112/transcriptr" title="Source Code" class="sidebar-tool px-1"><i class="bi bi-github"></i></a>
    <a href="" title="Download" id="sidebar-tool-dropdown-0" class="sidebar-tool dropdown-toggle px-1" data-bs-toggle="dropdown" aria-expanded="false"><i class="bi bi-download"></i></a>
    <ul class="dropdown-menu" aria-labelledby="sidebar-tool-dropdown-0">
        <li>
          <a class="dropdown-item sidebar-tools-main-item" href="./TranscriptR-manual.pdf">
            <i class="bi bi-bi-file-pdf pe-1"></i>
          Download PDF
          </a>
        </li>
        <li>
          <a class="dropdown-item sidebar-tools-main-item" href="./TranscriptR-manual.epub">
            <i class="bi bi-bi-journal pe-1"></i>
          Download ePub
          </a>
        </li>
    </ul>
    <a href="" title="Share" id="sidebar-tool-dropdown-1" class="sidebar-tool dropdown-toggle px-1" data-bs-toggle="dropdown" aria-expanded="false"><i class="bi bi-share"></i></a>
    <ul class="dropdown-menu" aria-labelledby="sidebar-tool-dropdown-1">
        <li>
          <a class="dropdown-item sidebar-tools-main-item" href="https://twitter.com/intent/tweet?url=|url|">
            <i class="bi bi-bi-twitter pe-1"></i>
          Twitter
          </a>
        </li>
        <li>
          <a class="dropdown-item sidebar-tools-main-item" href="https://www.facebook.com/sharer/sharer.php?u=|url|">
            <i class="bi bi-bi-facebook pe-1"></i>
          Facebook
          </a>
        </li>
    </ul>
</div>
    </div>
      </div>
      <div class="mt-2 flex-shrink-0 align-items-center">
        <div class="sidebar-search">
        <div id="quarto-search" class="" title="Search"></div>
        </div>
      </div>
    <div class="sidebar-menu-container"> 
    <ul class="list-unstyled mt-1">
        <li class="sidebar-item">
  <div class="sidebar-item-container"> 
  <a href="./index.html" class="sidebar-item-text sidebar-link">Acknowledgements</a>
  </div>
</li>
        <li class="sidebar-item sidebar-item-section">
      <div class="sidebar-item-container"> 
            <a href="./intro.html" class="sidebar-item-text sidebar-link">TranscriptR: Transcriptome Data Analysis Pipeline</a>
          <a class="sidebar-item-toggle text-start" data-bs-toggle="collapse" data-bs-target="#quarto-sidebar-section-1" aria-expanded="true">
            <i class="bi bi-chevron-right ms-2"></i>
          </a> 
      </div>
      <ul id="quarto-sidebar-section-1" class="collapse list-unstyled sidebar-section depth1 show">  
          <li class="sidebar-item">
  <div class="sidebar-item-container"> 
  <a href="./quality.html" class="sidebar-item-text sidebar-link"><span class="chapter-number">1</span>&nbsp; <span class="chapter-title">Quality Analysis</span></a>
  </div>
</li>
          <li class="sidebar-item">
  <div class="sidebar-item-container"> 
  <a href="./transcriptome.html" class="sidebar-item-text sidebar-link active"><span class="chapter-number">2</span>&nbsp; <span class="chapter-title">Transcriptome Analysis</span></a>
  </div>
</li>
          <li class="sidebar-item">
  <div class="sidebar-item-container"> 
  <a href="./igv.html" class="sidebar-item-text sidebar-link"><span class="chapter-number">3</span>&nbsp; <span class="chapter-title">Integrative Genome Viewer</span></a>
  </div>
</li>
          <li class="sidebar-item">
  <div class="sidebar-item-container"> 
  <a href="./Multi_D_Analysis.html" class="sidebar-item-text sidebar-link"><span class="chapter-number">4</span>&nbsp; <span class="chapter-title">Dimensional Analysis</span></a>
  </div>
</li>
          <li class="sidebar-item">
  <div class="sidebar-item-container"> 
  <a href="./Volcano_plot.html" class="sidebar-item-text sidebar-link"><span class="chapter-number">5</span>&nbsp; <span class="chapter-title">Volcano plot</span></a>
  </div>
</li>
          <li class="sidebar-item">
  <div class="sidebar-item-container"> 
  <a href="./heatmap.html" class="sidebar-item-text sidebar-link"><span class="chapter-number">6</span>&nbsp; <span class="chapter-title">HeatMap</span></a>
  </div>
</li>
      </ul>
  </li>
        <li class="sidebar-item sidebar-item-section">
      <div class="sidebar-item-container"> 
            <a href="./assco.html" class="sidebar-item-text sidebar-link">Enrichment Analysis</a>
          <a class="sidebar-item-toggle text-start" data-bs-toggle="collapse" data-bs-target="#quarto-sidebar-section-2" aria-expanded="true">
            <i class="bi bi-chevron-right ms-2"></i>
          </a> 
      </div>
      <ul id="quarto-sidebar-section-2" class="collapse list-unstyled sidebar-section depth1 show">  
          <li class="sidebar-item">
  <div class="sidebar-item-container"> 
  <a href="./go_analysis.html" class="sidebar-item-text sidebar-link"><span class="chapter-number">7</span>&nbsp; <span class="chapter-title">Gene Ontology (GO) enrichment analysis</span></a>
  </div>
</li>
          <li class="sidebar-item">
  <div class="sidebar-item-container"> 
  <a href="./pathway_analysis.html" class="sidebar-item-text sidebar-link"><span class="chapter-number">8</span>&nbsp; <span class="chapter-title">Pathway enrichment analysis</span></a>
  </div>
</li>
      </ul>
  </li>
        <li class="sidebar-item sidebar-item-section">
      <div class="sidebar-item-container"> 
            <a href="./setana.html" class="sidebar-item-text sidebar-link">Set Analysis</a>
          <a class="sidebar-item-toggle text-start" data-bs-toggle="collapse" data-bs-target="#quarto-sidebar-section-3" aria-expanded="true">
            <i class="bi bi-chevron-right ms-2"></i>
          </a> 
      </div>
      <ul id="quarto-sidebar-section-3" class="collapse list-unstyled sidebar-section depth1 show">  
          <li class="sidebar-item">
  <div class="sidebar-item-container"> 
  <a href="./venn_analysis.html" class="sidebar-item-text sidebar-link"><span class="chapter-number">9</span>&nbsp; <span class="chapter-title">Venn analysis</span></a>
  </div>
</li>
          <li class="sidebar-item">
  <div class="sidebar-item-container"> 
  <a href="./upset_plot.html" class="sidebar-item-text sidebar-link"><span class="chapter-number">10</span>&nbsp; <span class="chapter-title">UpSet Plots</span></a>
  </div>
</li>
      </ul>
  </li>
        <li class="sidebar-item sidebar-item-section">
      <div class="sidebar-item-container"> 
            <a class="sidebar-item-text sidebar-link text-start" data-bs-toggle="collapse" data-bs-target="#quarto-sidebar-section-4" aria-expanded="true">Appendices</a>
          <a class="sidebar-item-toggle text-start" data-bs-toggle="collapse" data-bs-target="#quarto-sidebar-section-4" aria-expanded="true">
            <i class="bi bi-chevron-right ms-2"></i>
          </a> 
      </div>
      <ul id="quarto-sidebar-section-4" class="collapse list-unstyled sidebar-section depth1 show">  
          <li class="sidebar-item">
  <div class="sidebar-item-container"> 
  <a href="./appendix1.html" class="sidebar-item-text sidebar-link"><span class="chapter-number">A</span>&nbsp; <span class="chapter-title">Create the input zip file for <em>methylR</em></span></a>
  </div>
</li>
          <li class="sidebar-item">
  <div class="sidebar-item-container"> 
  <a href="./appendix2.html" class="sidebar-item-text sidebar-link"><span class="chapter-number">B</span>&nbsp; <span class="chapter-title">Use of Docker Container</span></a>
  </div>
</li>
          <li class="sidebar-item">
  <div class="sidebar-item-container"> 
  <a href="./appendix3.html" class="sidebar-item-text sidebar-link"><span class="chapter-number">C</span>&nbsp; <span class="chapter-title">Convert DMCs table to BED</span></a>
  </div>
</li>
          <li class="sidebar-item">
  <div class="sidebar-item-container"> 
  <a href="./appendix4.html" class="sidebar-item-text sidebar-link"><span class="chapter-number">D</span>&nbsp; <span class="chapter-title">Time calculation</span></a>
  </div>
</li>
          <li class="sidebar-item">
  <div class="sidebar-item-container"> 
  <a href="./appendix5.html" class="sidebar-item-text sidebar-link"><span class="chapter-number">E</span>&nbsp; <span class="chapter-title">FAQs/Troubleshooting</span></a>
  </div>
</li>
          <li class="sidebar-item">
  <div class="sidebar-item-container"> 
  <a href="./references.html" class="sidebar-item-text sidebar-link">References</a>
  </div>
</li>
      </ul>
  </li>
    </ul>
    </div>
</nav>
<!-- margin-sidebar -->
    <div id="quarto-margin-sidebar" class="sidebar margin-sidebar"><div class="quarto-margin-header"><div class="margin-header-item">
<p><img src="_images/logo-final.png" class="img-fluid"></p>
</div></div>
        <nav id="TOC" role="doc-toc" class="toc-active">
    <h2 id="toc-title">Table of contents</h2>
   
  <ul>
  <li><a href="#data-input-and-parameters-settings" id="toc-data-input-and-parameters-settings" class="nav-link active" data-scroll-target="#data-input-and-parameters-settings"><span class="toc-section-number">2.1</span>  Data input and parameters settings</a>
  <ul class="collapse">
  <li><a href="#fastq-upload" id="toc-fastq-upload" class="nav-link" data-scroll-target="#fastq-upload"><span class="toc-section-number">2.1.1</span>  Fastq upload</a></li>
  <li><a href="#sec-sampleinfo" id="toc-sec-sampleinfo" class="nav-link" data-scroll-target="#sec-sampleinfo"><span class="toc-section-number">2.1.2</span>  Sample info</a></li>
  <li><a href="#parameters-settings" id="toc-parameters-settings" class="nav-link" data-scroll-target="#parameters-settings"><span class="toc-section-number">2.1.3</span>  Parameters settings</a></li>
  </ul></li>
  <li><a href="#visualization-multiqc" id="toc-visualization-multiqc" class="nav-link" data-scroll-target="#visualization-multiqc"><span class="toc-section-number">2.2</span>  Visualization: MultiQC</a></li>
  <li><a href="#visualization-plots" id="toc-visualization-plots" class="nav-link" data-scroll-target="#visualization-plots"><span class="toc-section-number">2.3</span>  Visualization: Plots</a></li>
  <li><a href="#visualization-deg-table" id="toc-visualization-deg-table" class="nav-link" data-scroll-target="#visualization-deg-table"><span class="toc-section-number">2.4</span>  Visualization: DEG table</a></li>
  <li><a href="#visualization-intersectde" id="toc-visualization-intersectde" class="nav-link" data-scroll-target="#visualization-intersectde"><span class="toc-section-number">2.5</span>  Visualization: InterSectDE</a></li>
  <li><a href="#session-logs" id="toc-session-logs" class="nav-link" data-scroll-target="#session-logs"><span class="toc-section-number">2.6</span>  Session &amp; logs</a></li>
  <li><a href="#download-results" id="toc-download-results" class="nav-link" data-scroll-target="#download-results"><span class="toc-section-number">2.7</span>  Download Results</a></li>
  </ul>
<div class="toc-actions"><div><i class="bi bi-github"></i></div><div class="action-links"><p><a href="https://github.com/JD2112/transcriptr/issues/new" class="toc-action">Report an issue</a></p></div></div></nav>
    </div>
<!-- main -->
<main class="content" id="quarto-document-content">

<header id="title-block-header" class="quarto-title-block default">
<div class="quarto-title">
<h1 class="title"><span id="sec-transcript" class="quarto-section-identifier d-none d-lg-block"><span class="chapter-number">2</span>&nbsp; <span class="chapter-title">Transcriptome Analysis</span></span></h1>
</div>



<div class="quarto-title-meta">

    
  
    
  </div>
  

</header>

<div class="callout-caution callout callout-style-default callout-captioned">
<div class="callout-header d-flex align-content-center">
<div class="callout-icon-container">
<i class="callout-icon"></i>
</div>
<div class="callout-caption-container flex-fill">
Danger
</div>
</div>
<div class="callout-body-container callout-body">
<p><strong>PLEASE NOTE</strong> WE INCLUDED THE POSSIBILITY OF THE HUMAN TRANSCRIPTOME DATA ANALYSIS FROM THE SEQUENCE DATA. NONE OF THE OWNER OF THE CODE OR DATA CENTER IS RESPONSIBLE FOR HANDLING THE HUMAN DATA. IT IS UPTO THE USER IF THEY WANT TO USE THE ONLINE VERSION FOR HUMAN DATA ANALYSIS.</p>
</div>
</div>
<p>Transcriptome analysis has divided in several parts for the data upload, analysis and visualizing the results.</p>
<section id="data-input-and-parameters-settings" class="level2" data-number="2.1">
<h2 data-number="2.1" class="anchored" data-anchor-id="data-input-and-parameters-settings"><span class="header-section-number">2.1</span> Data input and parameters settings</h2>
<p>In this section, for better user accessibility, we divided this into 3 sub-parts -</p>
<div id="fig-trans1" class="quarto-figure quarto-figure-center anchored">
<figure class="figure">
<p><img src="_images/transcriptome/transcript1.png" class="img-fluid figure-img"></p>
<p></p><figcaption class="figure-caption">Figure&nbsp;2.1: Fastq upload</figcaption><p></p>
</figure>
</div>
<section id="fastq-upload" class="level3" data-number="2.1.1">
<h3 data-number="2.1.1" class="anchored" data-anchor-id="fastq-upload"><span class="header-section-number">2.1.1</span> Fastq upload</h3>
<ol type="1">
<li>If the user performed the <strong><em>Quality analysis</em></strong> (<a href="quality.html"><span>Chapter&nbsp;1</span></a>) already, they don’t need to upload the data again, they can use <strong><em>‘continue from Quality Analysis’</em></strong>.</li>
<li>The user can also upload the data directly by selecting <strong><em>‘upload data’</em></strong> option.</li>
</ol>
<div class="callout-note callout callout-style-default callout-captioned">
<div class="callout-header d-flex align-content-center">
<div class="callout-icon-container">
<i class="callout-icon"></i>
</div>
<div class="callout-caption-container flex-fill">
Note
</div>
</div>
<div class="callout-body-container callout-body">
<p>Please note it will take some time to upload the data and once the data uploaded, the list will be displayed in the <strong>Data file viewer</strong> (see <a href="quality.html#fig-qual2">Figure&nbsp;<span>1.2</span></a> <a href="#fig-trans1">Figure&nbsp;<span>2.1</span></a>).</p>
</div>
</div>
</section>
<section id="sec-sampleinfo" class="level3" data-number="2.1.2">
<h3 data-number="2.1.2" class="anchored" data-anchor-id="sec-sampleinfo"><span class="header-section-number">2.1.2</span> Sample info</h3>
<p>In the <strong><em>Sample info</em></strong> section, users need to upload the ‘sample sheet/meta data file’ that includes atleast two columns (‘sample name’ and ‘sample condition’) (see <a href="#fig-trans2">Figure&nbsp;<span>2.2</span></a>). For Differential Gene Expression (DEG) analysis, user need to provide the comparison conditions as showed in <a href="#fig-trans2">Figure&nbsp;<span>2.2</span></a>.</p>
<div id="fig-trans2" class="quarto-figure quarto-figure-center anchored">
<figure class="figure">
<p><img src="_images/transcriptome/transcriptr3.png" class="img-fluid figure-img"></p>
<p></p><figcaption class="figure-caption">Figure&nbsp;2.2: Sample information</figcaption><p></p>
</figure>
</div>
<div class="callout-tip callout callout-style-default callout-captioned">
<div class="callout-header d-flex align-content-center">
<div class="callout-icon-container">
<i class="callout-icon"></i>
</div>
<div class="callout-caption-container flex-fill">
Tip
</div>
</div>
<div class="callout-body-container callout-body">
<ol type="1">
<li>After uploading the files, all files will be appear in the browser window (see <a href="#fig-trans2">Figure&nbsp;<span>2.2</span></a>), please check and verify everything is properly uploaded.</li>
<li>User can upload any file format, such as text (.txt), CSV (.csv), Excel (.xls, .xlsx).</li>
</ol>
</div>
</div>
</section>
<section id="parameters-settings" class="level3" data-number="2.1.3">
<h3 data-number="2.1.3" class="anchored" data-anchor-id="parameters-settings"><span class="header-section-number">2.1.3</span> Parameters settings</h3>
<p>Selecting the right parameters for the analysis is the key to the successful run of <strong>TranscriptR</strong>. The <strong>parameters settings</strong> (see <a href="#fig-trans3">Figure&nbsp;<span>2.3</span></a>) are sub-divided for better understanding and view -</p>
<div id="fig-trans3" class="quarto-figure quarto-figure-center anchored">
<figure class="figure">
<p><img src="_images/transcriptome/transcriptr4.png" class="img-fluid figure-img"></p>
<p></p><figcaption class="figure-caption">Figure&nbsp;2.3: Parameters settings</figcaption><p></p>
</figure>
</div>
<section id="species-and-database" class="level4" data-number="2.1.3.1">
<h4 data-number="2.1.3.1" class="anchored" data-anchor-id="species-and-database"><span class="header-section-number">2.1.3.1</span> Species and database</h4>
<ol type="1">
<li><strong>Select Species</strong>: The user need to select the <em>species</em> from the species list. Here we provided more than 350 species presented in the ‘biomart’ for the analysis. Note that all other species than ‘Human’, ‘Mouse’ and ‘Rat’ come with the <strong>scientific name</strong> as provided in <a href="http://www.ensembl.org/info/about/species.html" target="_blank">‘Ensembl’</a>. The default is ‘Human’.</li>
<li><strong>Select release version</strong>: The default is <em>latest</em>. The user can choose other database version as provided by <a href="https://ftp.ensembl.org/pub/" target="_blank">Ensembl</a>. Please provide the number only in the ‘box’, example, if you want to use version ‘109’, provide only ‘109’ (see <a href="#fig-trans4">Figure&nbsp;<span>2.4</span></a>).</li>
</ol>
<div id="fig-trans4" class="quarto-figure quarto-figure-center anchored">
<figure class="figure">
<p><img src="_images/transcriptome/transcript5.png" class="img-fluid figure-img"></p>
<p></p><figcaption class="figure-caption">Figure&nbsp;2.4: Parameters: Species and database</figcaption><p></p>
</figure>
</div>
</section>
<section id="alignment-parameters" class="level4" data-number="2.1.3.2">
<h4 data-number="2.1.3.2" class="anchored" data-anchor-id="alignment-parameters"><span class="header-section-number">2.1.3.2</span> Alignment parameters</h4>
<ol type="1">
<li><strong>Choose R1/R2 suffix</strong>: For <em>paired-end</em> analysis, please provide the R1 and R2 suffixes.</li>
<li><strong>Overhang</strong>: The default is 100.</li>
<li><strong>Choose sequence type</strong>: Choose whether the sequence is <em>stranded</em>, <em>unstranded</em> or <em>reverse-stranded</em>.</li>
<li><strong>Choose ID</strong>: Select display ID for <em>gene id</em> or <em>transcriptome id</em>. Default is <em>gene ID</em>.</li>
</ol>
<div id="fig-trans5" class="quarto-figure quarto-figure-center anchored">
<figure class="figure">
<p><img src="_images/transcriptome/transcript6.png" class="img-fluid figure-img"></p>
<p></p><figcaption class="figure-caption">Figure&nbsp;2.5: Parameters: Alignment</figcaption><p></p>
</figure>
</div>
</section>
<section id="deg-filtration" class="level4" data-number="2.1.3.3">
<h4 data-number="2.1.3.3" class="anchored" data-anchor-id="deg-filtration"><span class="header-section-number">2.1.3.3</span> DEG filtration</h4>
<p>As <strong>‘TranscriptR’</strong> provides the differential gene expression (DEG) analysis result, user need to provide the required filtration for the analysis.<br>
1. <strong>Enter CPM value</strong>: The default is 1.<br>
2. <strong>Number of samples</strong>: The default is 3.<br>
3. <strong>Choose logFC value</strong>: The default is 1.5.<br>
4. <strong>Choose FDR value</strong>: The default is 0.05.</p>
<div id="fig-trans6" class="quarto-figure quarto-figure-center anchored">
<figure class="figure">
<p><img src="_images/transcriptome/transcript7.png" class="img-fluid figure-img"></p>
<p></p><figcaption class="figure-caption">Figure&nbsp;2.6: Parameters: DEF filtration</figcaption><p></p>
</figure>
</div>
</section>
</section>
</section>
<section id="visualization-multiqc" class="level2" data-number="2.2">
<h2 data-number="2.2" class="anchored" data-anchor-id="visualization-multiqc"><span class="header-section-number">2.2</span> Visualization: MultiQC</h2>
<p>Please see <a href="quality.html"><span>Chapter&nbsp;1</span></a> for details.</p>
</section>
<section id="visualization-plots" class="level2" data-number="2.3">
<h2 data-number="2.3" class="anchored" data-anchor-id="visualization-plots"><span class="header-section-number">2.3</span> Visualization: Plots</h2>
<p>Since <strong>TranscriptR</strong> provides differential gene expression (DEG) analysis by <strong>three</strong> different pipelines, namely <em>EdgeR</em>, <em>DESeq2</em> and <em>limma</em>, all plots are divided into these three results.</p>
<div id="fig-trans7" class="quarto-figure quarto-figure-center anchored">
<figure class="figure">
<p><img src="_images/transcriptome/transcript8.png" class="img-fluid figure-img"></p>
<p></p><figcaption class="figure-caption">Figure&nbsp;2.7: Visualization: plots</figcaption><p></p>
</figure>
</div>
<div class="callout-note callout callout-style-default callout-captioned">
<div class="callout-header d-flex align-content-center">
<div class="callout-icon-container">
<i class="callout-icon"></i>
</div>
<div class="callout-caption-container flex-fill">
Note
</div>
</div>
<div class="callout-body-container callout-body">
<ol type="1">
<li>The plots will be displayed as <a href="#fig-trans7">Figure&nbsp;<span>2.7</span></a> in a PDF format. All of these can be downloaded when the user download the results from the <em>‘Download results’</em> tab.</li>
<li>Please note, if the plots are not loaded after the <em>job</em> completed and clicked on ‘show BoxPlot/Show MDS Plot’ button, refresh the browser and it will appear in the box.</li>
</ol>
</div>
</div>
</section>
<section id="visualization-deg-table" class="level2" data-number="2.4">
<h2 data-number="2.4" class="anchored" data-anchor-id="visualization-deg-table"><span class="header-section-number">2.4</span> Visualization: DEG table</h2>
<p>We provided differential gene expression (DEG) analysis result as a table format (see <a href="#fig-trans8">Figure&nbsp;<span>2.8</span></a>). To display the table, user need to - &nbsp; 1. choose the analysis type (as <strong>TranscriptR</strong> provides three different pipeline results), such as <em>EdgeR analysis</em>, <em>DESeq analysis</em> or <em>limma analysis</em> and<br>
2. select DEG table from the drop-down list. The list is generated from the information provided by the <em>comparison table</em>.</p>
<div id="fig-trans8" class="quarto-figure quarto-figure-center anchored">
<figure class="figure">
<p><img src="_images/transcriptome/transcript9.png" class="img-fluid figure-img"></p>
<p></p><figcaption class="figure-caption">Figure&nbsp;2.8: Visualization: DEG table</figcaption><p></p>
</figure>
</div>
<p>In the DEG table, we added information for -<br>
1. <strong>ID</strong>: ENSEMBL <em>gene ID</em> or <em>transcript ID</em><br>
2. <strong>logFC</strong>: log fold change value for each ID. Red color shows the higher expression and blue represents the low expression (same corresponds to <em>gene symbol</em>).<br>
3. <strong>logCPM</strong>: log CPM value provided<br>
4. <strong>P_value</strong>: <em>p</em>-value provided for each ID.<br>
5. <strong>FDR</strong>: FDR provided for each ID.<br>
6. <strong>Gene Symbol</strong>: official gene symbol added in the table. Coloured according to the <em>logFC</em> value.<br>
7. <strong>Ensembl link</strong>: we added external link for ensembl database to view further information on the ID.<br>
8. <strong>GeneCard link</strong>: we added genecard database link for more information on the ID.<br>
9. <strong>Gene details</strong>: Clicking on <em>’show details</em> button will result to display the boxplot of CPM data for each group (<strong>see the note below</strong>) and the gene information from <strong>HGNC</strong> (human only).<br>
10. <strong>Chr</strong>: Chromosomal location<br>
11. <strong>Start</strong>: Start position of the chromosome<br>
12. <strong>End</strong>: End position of the chromosome<br>
13. <strong>Strand</strong>: chromosomal strand<br>
14. <strong>Types</strong>: types of genes/transcript.</p>
<div class="callout-important callout callout-style-default callout-captioned">
<div class="callout-header d-flex align-content-center">
<div class="callout-icon-container">
<i class="callout-icon"></i>
</div>
<div class="callout-caption-container flex-fill">
Important
</div>
</div>
<div class="callout-body-container callout-body">
<p>Please note, we already mentioned that the user may need to reload/refresh the browser after successful run to visualize the results. In this process, the uploaded data table (<strong>mentioned in <a href="#sec-sampleinfo"><span>Section&nbsp;2.1.2</span></a></strong>) will be lost. To view the CPM boxplot, the user need the upload the sample information data tables again and click on <em>‘Show details’</em> on the DEG table.</p>
</div>
</div>
</section>
<section id="visualization-intersectde" class="level2" data-number="2.5">
<h2 data-number="2.5" class="anchored" data-anchor-id="visualization-intersectde"><span class="header-section-number">2.5</span> Visualization: InterSectDE</h2>
<p>Further, we implemented intersection function for the above-mentioned three types of DEG analyses in the <em>’InterSectDEG</em> tab.<br>
1. To visualize the result, the user needs to select the <em>same result</em> table (eg. groupAvsgroupB) for all three analyses. <strong>Please remember selecting different table may results incorrect table and figure.<br>
2. After selecting same result table files from all three analyses, click to <em>‘show intersect result’</em>. Both intersect table and Venn diagram will appear in respective boxes.<br>
3. To download the full table, check to show </strong>all** entries (see <a href="#fig-trans9">Figure&nbsp;<span>2.9</span></a>).<br>
4. The Venn diagram can be downloaded in different file format with respective height, width or resolution (see <a href="#fig-trans9">Figure&nbsp;<span>2.9</span></a>).</p>
<div id="fig-trans9" class="quarto-figure quarto-figure-center anchored">
<figure class="figure">
<p><img src="_images/transcriptome/transcript10.png" class="img-fluid figure-img"></p>
<p></p><figcaption class="figure-caption">Figure&nbsp;2.9: InterSect result</figcaption><p></p>
</figure>
</div>
<p>Below is the details of <em>‘InterSect DEG’</em> table result (see <a href="#fig-trans9">Figure&nbsp;<span>2.9</span></a>) -<br>
1. <strong>id</strong><br>
2. <strong>name</strong><br>
3. <strong>chr</strong><br>
4. <strong>start</strong><br>
5. <strong>end</strong><br>
6. <strong>strand</strong><br>
7. <strong>biotype</strong><br>
8. <strong>edgeR_logFC</strong><br>
9. <strong>edgeR_logCPM</strong><br>
10. <strong>edgeR_PValue</strong><br>
11. <strong>edgeR_FDR</strong><br>
12. <strong>edgeR_baseMean</strong><br>
13. <strong>DESeq2_log2FoldChange</strong><br>
14. <strong>DESeq2_lfcSE</strong><br>
15. <strong>DESeq2_stat</strong><br>
16. <strong>DESeq2_pvalue</strong><br>
17. <strong>DESeq2_padj</strong><br>
18. <strong>limma_logFC</strong><br>
19. <strong>limma_AveExpr</strong><br>
20. <strong>limma_t</strong><br>
21. <strong>limma_P.Value</strong><br>
22. <strong>limma_adj.P.Val</strong><br>
23. <strong>limma_B</strong></p>
</section>
<section id="session-logs" class="level2" data-number="2.6">
<h2 data-number="2.6" class="anchored" data-anchor-id="session-logs"><span class="header-section-number">2.6</span> Session &amp; logs</h2>
<p>We included the next tab to show <em>‘GTF information’</em>, <em>‘Reference information’</em>, <em>‘Shiny logs’</em> and <em>‘Session information’</em> for the pipeline run.</p>
</section>
<section id="download-results" class="level2" data-number="2.7">
<h2 data-number="2.7" class="anchored" data-anchor-id="download-results"><span class="header-section-number">2.7</span> Download Results</h2>
<p>In the <em>‘Download results’</em> tab, user can download all results performed and saved during the <strong>Transcriptome Analysis</strong> run in a zip folder. This also contains the snakemake and session logs.</p>


</section>

<p>Copyright © 2023-2024. Massimiliano Volpe and Jyotirmoy Das.</p></main> <!-- /main -->
<script id="quarto-html-after-body" type="application/javascript">
window.document.addEventListener("DOMContentLoaded", function (event) {
  const toggleBodyColorMode = (bsSheetEl) => {
    const mode = bsSheetEl.getAttribute("data-mode");
    const bodyEl = window.document.querySelector("body");
    if (mode === "dark") {
      bodyEl.classList.add("quarto-dark");
      bodyEl.classList.remove("quarto-light");
    } else {
      bodyEl.classList.add("quarto-light");
      bodyEl.classList.remove("quarto-dark");
    }
  }
  const toggleBodyColorPrimary = () => {
    const bsSheetEl = window.document.querySelector("link#quarto-bootstrap");
    if (bsSheetEl) {
      toggleBodyColorMode(bsSheetEl);
    }
  }
  toggleBodyColorPrimary();  
  const icon = "";
  const anchorJS = new window.AnchorJS();
  anchorJS.options = {
    placement: 'right',
    icon: icon
  };
  anchorJS.add('.anchored');
  const clipboard = new window.ClipboardJS('.code-copy-button', {
    target: function(trigger) {
      return trigger.previousElementSibling;
    }
  });
  clipboard.on('success', function(e) {
    // button target
    const button = e.trigger;
    // don't keep focus
    button.blur();
    // flash "checked"
    button.classList.add('code-copy-button-checked');
    var currentTitle = button.getAttribute("title");
    button.setAttribute("title", "Copied!");
    let tooltip;
    if (window.bootstrap) {
      button.setAttribute("data-bs-toggle", "tooltip");
      button.setAttribute("data-bs-placement", "left");
      button.setAttribute("data-bs-title", "Copied!");
      tooltip = new bootstrap.Tooltip(button, 
        { trigger: "manual", 
          customClass: "code-copy-button-tooltip",
          offset: [0, -8]});
      tooltip.show();    
    }
    setTimeout(function() {
      if (tooltip) {
        tooltip.hide();
        button.removeAttribute("data-bs-title");
        button.removeAttribute("data-bs-toggle");
        button.removeAttribute("data-bs-placement");
      }
      button.setAttribute("title", currentTitle);
      button.classList.remove('code-copy-button-checked');
    }, 1000);
    // clear code selection
    e.clearSelection();
  });
  function tippyHover(el, contentFn) {
    const config = {
      allowHTML: true,
      content: contentFn,
      maxWidth: 500,
      delay: 100,
      arrow: false,
      appendTo: function(el) {
          return el.parentElement;
      },
      interactive: true,
      interactiveBorder: 10,
      theme: 'quarto',
      placement: 'bottom-start'
    };
    window.tippy(el, config); 
  }
  const noterefs = window.document.querySelectorAll('a[role="doc-noteref"]');
  for (var i=0; i<noterefs.length; i++) {
    const ref = noterefs[i];
    tippyHover(ref, function() {
      // use id or data attribute instead here
      let href = ref.getAttribute('data-footnote-href') || ref.getAttribute('href');
      try { href = new URL(href).hash; } catch {}
      const id = href.replace(/^#\/?/, "");
      const note = window.document.getElementById(id);
      return note.innerHTML;
    });
  }
  const findCites = (el) => {
    const parentEl = el.parentElement;
    if (parentEl) {
      const cites = parentEl.dataset.cites;
      if (cites) {
        return {
          el,
          cites: cites.split(' ')
        };
      } else {
        return findCites(el.parentElement)
      }
    } else {
      return undefined;
    }
  };
  var bibliorefs = window.document.querySelectorAll('a[role="doc-biblioref"]');
  for (var i=0; i<bibliorefs.length; i++) {
    const ref = bibliorefs[i];
    const citeInfo = findCites(ref);
    if (citeInfo) {
      tippyHover(citeInfo.el, function() {
        var popup = window.document.createElement('div');
        citeInfo.cites.forEach(function(cite) {
          var citeDiv = window.document.createElement('div');
          citeDiv.classList.add('hanging-indent');
          citeDiv.classList.add('csl-entry');
          var biblioDiv = window.document.getElementById('ref-' + cite);
          if (biblioDiv) {
            citeDiv.innerHTML = biblioDiv.innerHTML;
          }
          popup.appendChild(citeDiv);
        });
        return popup.innerHTML;
      });
    }
  }
});
</script>
<nav class="page-navigation">
  <div class="nav-page nav-page-previous">
      <a href="./quality.html" class="pagination-link">
        <i class="bi bi-arrow-left-short"></i> <span class="nav-page-text"><span class="chapter-number">1</span>&nbsp; <span class="chapter-title">Quality Analysis</span></span>
      </a>          
  </div>
  <div class="nav-page nav-page-next">
      <a href="./igv.html" class="pagination-link">
        <span class="nav-page-text"><span class="chapter-number">3</span>&nbsp; <span class="chapter-title">Integrative Genome Viewer</span></span> <i class="bi bi-arrow-right-short"></i>
      </a>
  </div>
</nav>
</div> <!-- /content -->



</body></html>